test/extensions/transport_sockets/http_11_proxy/connect_test.cc

```diff
@@ -0,0 +1,353 @@
+        StreamInfo::FilterState::StateType::ReadOnly,
+        StreamInfo::FilterState::LifeSpan::FilterChain);
+  }
+
+  Network::TransportSocketOptionsConstSharedPtr options_;
+  NiceMock<Network::MockTransportSocket>* inner_socket_;
+  NiceMock<Network::MockIoHandle> io_handle_;
+  std::unique_ptr<UpstreamHttp11ConnectSocket> connect_socket_;
+  NiceMock<Network::MockTransportSocketCallbacks> transport_callbacks_;
+  Buffer::OwnedImpl connect_data_{"CONNECT www.foo.com:443 HTTP/1.1\r\n\r\n"};
+  Ssl::ConnectionInfoConstSharedPtr ssl_{
+      std::make_shared<NiceMock<Envoy::Ssl::MockConnectionInfo>>()};
+};
+
+// Test injects CONNECT  only once
+TEST_P(Http11ConnectTest, InjectesHeaderOnlyOnce) {
+  initialize();
+
+  EXPECT_CALL(io_handle_, write(BufferStringEqual(connect_data_.toString())))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) -> Api::IoCallUint64Result {
+        auto length = buffer.length();
+        buffer.drain(length);
+        return Api::IoCallUint64Result(length, Api::IoErrorPtr(nullptr, [](Api::IoError*) {}));
+      }));
+  auto msg = Buffer::OwnedImpl("initial data");
+
+  auto rc1 = connect_socket_->doWrite(msg, false);
+  // Only the connect will be written initially. All other writes should be
+  // buffered in the Network::Connection buffer until the connect has been
+  // processed.
+  EXPECT_EQ(connect_data_.length(), rc1.bytes_processed_);
+  auto rc2 = connect_socket_->doWrite(msg, false);
+  EXPECT_EQ(0, rc2.bytes_processed_);
+
+  EXPECT_CALL(*inner_socket_, onConnected());
+  connect_socket_->onConnected();
+}
+
+// Test the socket is a no-op if there's no header proto.
+TEST_P(Http11ConnectTest, NoInjectHeaderProtoAbsent) {
+  initialize(true);
+
+  EXPECT_CALL(io_handle_, write(_)).Times(0);
+  auto msg = Buffer::OwnedImpl("initial data");
+  auto msg2 = Buffer::OwnedImpl("new data");
+
+  {
+    InSequence s;
+    EXPECT_CALL(*inner_socket_, doWrite(BufferEqual(&msg), false))
+        .WillOnce(Return(Network::IoResult{Network::PostIoAction::KeepOpen, msg.length(), false}));
+    EXPECT_CALL(*inner_socket_, doWrite(BufferEqual(&msg2), false))
+        .WillOnce(Return(Network::IoResult{Network::PostIoAction::KeepOpen, msg2.length(), false}));
+  }
+
+  auto rc1 = connect_socket_->doWrite(msg, false);
+  EXPECT_EQ(msg.length(), rc1.bytes_processed_);
+  auto rc2 = connect_socket_->doWrite(msg2, false);
+  EXPECT_EQ(msg2.length(), rc2.bytes_processed_);
+
+  EXPECT_CALL(*inner_socket_, onConnected());
+  connect_socket_->onConnected();
+
+  // Make sure the response path is a no-op as well.
+  EXPECT_CALL(io_handle_, recv(_, _, _)).Times(0);
+  EXPECT_CALL(io_handle_, read(_, _)).Times(0);
+  Buffer::OwnedImpl buffer("");
+  connect_socket_->doRead(buffer);
+}
+
+TEST_P(Http11ConnectTest, NoInjectTlsAbsent) {
+  ssl_.reset();
+  initialize(false);
+
+  EXPECT_CALL(io_handle_, write(_)).Times(0);
+  auto msg = Buffer::OwnedImpl("initial data");
```

Comment:
How about `Buffer::OwnedImpl msg("initial data")` which is shorter and does not use auto.

Reply:
done